### PR TITLE
feat:implement loading student summary using nestjs

### PR DESCRIPTION
**Pull request by Alphajax**
**🟢 Add `deploy` label if you want to deploy this Pull Request to staging environment**

#### 🧑‍⚖️ Pull Request Naming Convention

- Title should follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)
- Do not put issue id in title
- Do not put WIP in title. Use [Draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/) functionality
- Consider to add `area:*` label(s)

* [x] I followed naming convention rules

---

#### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [x] Performance optimization
- [x] Refactoring
- [ ] Test Case
- [ ] Documentation update
- [ ] Other

#### 🔗 [Related issue link](https://github.com/rolling-scopes/rsschool-app/issues/2435)

_Describe the source of requirement, like related issue link._

#### 💡 Background and solution

_Describe the big picture of your changes here_
- implement new api in nestjs backend which clones functionality of getStudentSummary route from server
- fix incorrect tech check point view on profile page
#### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Database migration is added or not needed
- [x] Documentation is updated/provided or not needed
- [x] Changes are tested locally

### CODE REVIEW

File: nestjs/src/courses/course-students/course-students.controller.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    @Param('courseId') courseId: number,
+    @Param('githubId') githubId: string,
+    @Req() req: CurrentRequest,
+  ) {
+    let studentGithubId;
+    if (githubId === 'me') {
+      studentGithubId = req.user.githubId;
+    } else {
+      studentGithubId = githubId;
+    }
```
Comment (apalchys) [inline nestjs/src/courses/course-students/course-students.controller.ts:30]:
```suggestion
    const studentGithubId = githubId === 'me' ? req.user.githubId : githubId;
```

File: nestjs/src/courses/course-students/course-students.controller.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    const student = await this.courseStudentService.getStudentByGithubId(courseId, studentGithubId);
+
+    const [score, mentor] = await Promise.all([
+      this.courseStudentService.getStudentScore(student?.id || 0),
```
Comment (apalchys) [inline nestjs/src/courses/course-students/course-students.controller.ts:35]: `student?.` if no user throw an error. Don't need to hide it.

Comment (apalchys) [reply nestjs/src/courses/course-students/course-students.controller.ts:35]: e.g. do 

```typescript
if (student == null) {
   throw new NotFound(...)
}
```

File: nestjs/src/courses/course-students/course-students.service.ts [hunk 3]
```diff
@@ -0,0 +1,177 @@
+import { User } from '@entities/user';
+import { MentorBasic, StageInterviewFeedbackJson } from '@common/models';
```
Comment (apalchys) [inline nestjs/src/courses/course-students/course-students.service.ts:2]: Please do not import from `@common/` 
If you need a model, just copy to nest. We want to kill `@common`

Comment (Alphajax) [reply nestjs/src/courses/course-students/course-students.service.ts:2]: @valerydluski

File: nestjs/src/courses/course-students/course-students.service.ts [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+  mentorQuery() {
+    return this.mentorRepository.createQueryBuilder('mentor');
+  }
+
+  async getStudentByGithubId(courseId: number, githubId: string): Promise<Student | null> {
+    const record = await this.studentQuery()
+      .innerJoin('student.user', 'user')
+      .where('user.githubId = :githubId', { githubId })
+      .andWhere('student.courseId = :courseId', { courseId })
+      .getOne();
```
Comment (apalchys) [inline nestjs/src/courses/course-students/course-students.service.ts:35]: let's rewrite it and do like:

```typescript
this.studentRepository.findOne({
    where: { 
      courseId,
      user: { githubId },
    },
    relations: ['user']
  })
```

File: nestjs/src/courses/course-students/course-students.service.ts (hunk 3 above)
Comment (valerydluski) [reply nestjs/src/courses/course-students/course-students.service.ts:2]: @Alphajax When you asked me, I thought you meant Entity ([example](https://github.com/rolling-scopes/rsschool-app/blob/master/server/src/models/mentor.ts#L22)). We can confidently move the rest to Nest. I apologize for the misunderstanding, my bad.

Comment (Alphajax) [reply nestjs/src/courses/course-students/course-students.service.ts:2]: @valerydluski thank you fo explanation

File: nestjs/src/courses/course-schedule/course-schedule.service.ts [hunk 5]
```diff
@@ -304,7 +304,7 @@ export class CourseScheduleService {
         ...(technicalScreeningResults
           .find(task => task.courseTaskId === courseTaskId)
           ?.stageInterviewFeedbacks.map(feedback => JSON.parse(feedback.json))
-          .map((json: any) => json?.resume?.score ?? 0) ?? []),
+          .map((json: any) => (json?.resume?.score || json?.steps?.decision?.values?.finalScore) ?? 0) ?? []),
```
Comment (AlreadyBored) [inline nestjs/src/courses/course-schedule/course-schedule.service.ts:307]: Let's split it a bit and make more readable  
```suggestion
          .map((json: any) => {
    const resumeScore = json?.resume?.score;
    const decisionScore = json?.steps?.decision?.values?.finalScore;
    return resumeScore ?? decisionScore ?? 0;
});
```

File: nestjs/src/courses/course-students/course-students.controller.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+      student?.mentorId ? await this.courseStudentService.getMentorWithContacts(student.mentorId) : null,
+    ]);
+
+    return new StudentSummaryDto({
+      totalScore: score?.totalScore,
+      results: score?.results,
+      rank: score?.rank,
+      isActive: !student?.isExpelled && !student?.isFailed,
+      mentor,
+      repository: student?.repository || null,
```
Comment (AlreadyBored) [inline nestjs/src/courses/course-students/course-students.controller.ts:45]:
```suggestion
      repository: student?.repository ?? null,
```

File: nestjs/src/courses/course-students/course-students.controller.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  @ApiOkResponse({
+    type: StudentSummaryDto,
+  })
+  @ApiOperation({ operationId: 'getStudentSummary' })
+  public async getStudentSummary(
+    @Param('courseId') courseId: number,
+    @Param('githubId') githubId: string,
+    @Req() req: CurrentRequest,
+  ) {
+    const studentGithubId = githubId === 'me' ? req.user.githubId : githubId;
```
Comment (AlreadyBored) [inline nestjs/src/courses/course-students/course-students.controller.ts:25]: To be honest, I don't like this usage of `githubId`, such profile exist in github (https://github.com/me). Let's replace it with separated param that explicitly means that we should use user's github id

File: nestjs/src/courses/course-students/course-students.service.ts [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+  }
+}
+
+const getStageInterviewRating = (stageInterviews: StageInterview[]) => {
+  const [lastInterview] = stageInterviews
+    .filter((interview: StageInterview) => interview.isCompleted)
+    .map(({ stageInterviewFeedbacks, score }: StageInterview) =>
+      stageInterviewFeedbacks.map((feedback: StageInterviewFeedback) => ({
+        date: feedback.updatedDate,
+        // interviews in new template should have score precalculated
```
Comment (AlreadyBored) [inline nestjs/src/courses/course-students/course-students.service.ts:127]:
```suggestion
        // interviews in new template should have precalculated score
```

File: nestjs/src/courses/course-students/course-students.service.ts [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+        rating: score ?? getInterviewRatings(JSON.parse(feedback.json) as StageInterviewFeedbackJson).rating,
+      })),
+    )
+    .reduce((acc, cur) => acc.concat(cur), [])
+    .sort((a, b) => new Date(b.date).getTime() - new Date(a.date).getTime());
+
+  return lastInterview && lastInterview.rating !== undefined ? lastInterview.rating : null;
```
Comment (AlreadyBored) [inline nestjs/src/courses/course-students/course-students.service.ts:134]: Let's shorten it
```suggestion
  return lastInterview?.rating ?? null;
```

File: nestjs/src/courses/course-students/course-students.service.ts [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+  const user = (mentor.user as User)!;
+  return {
+    isActive: !mentor.isExpelled,
+    name: createName(user),
+    id: mentor.id,
+    githubId: user.githubId,
+    students: mentor.students ? mentor.students.filter(s => !s.isExpelled && !s.isFailed).map(s => ({ id: s.id })) : [],
+    cityName: user.cityName ?? '',
+    countryName: user.countryName ?? '',
+  };
```
Comment (AlreadyBored) [inline nestjs/src/courses/course-students/course-students.service.ts:167]: Type of mentor.user is User without this typecast
```suggestion
  return {
    isActive: !mentor.isExpelled,
    name: createName(mentor.user),
    id: mentor.id,
    githubId: user.githubId,
    students: mentor.students ? mentor.students.filter(s => !s.isExpelled && !s.isFailed).map(s => ({ id: s.id })) : [],
    cityName: user.cityName ?? '',
    countryName: user.countryName ?? '',
  };
```

File: nestjs/src/courses/course-students/types/stage-interview-feedback-json.ts [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+export type EnglishLevel = 'a0' | 'a1' | 'a1+' | 'a2' | 'a2+' | 'b1' | 'b1+' | 'b2' | 'b2+' | 'c1' | 'c1+' | 'c2';
+
+export interface StageInterviewFeedbackJson extends StageInterviewFeedback {
+  english: {
+    levelStudentOpinion: EnglishLevel | null;
+    levelMentorOpinion: EnglishLevel | null;
+    whereAndWhenLearned: string | null;
+    comment: string | null;
+  };
+}
```
Comment (AlreadyBored) [inline nestjs/src/courses/course-students/types/stage-interview-feedback-json.ts:58]: We have all these models in `common/models`, why do you duplicate it here?

**Review by apalchys — APPROVED** [review]

**Pull request merged**